src/Kiota.Builder/Writers/Shell/ShellCodeMethodWriter.cs

```diff
@@ -54,19 +54,22 @@ protected override void WriteCommandBuilderBody(CodeMethod codeElement, RequestP
                         var builderMethods = (codeElement.OriginalIndexer.ReturnType as CodeType).TypeDefinition.GetChildElements(true).OfType<CodeMethod>()
                             .Where(m => m.IsOfKind(CodeMethodKind.CommandBuilder))
                             .OrderBy(m => m.Name);
-                        // Filter out list builder commands. They contain the item builder commands already
-                        var itemBuilderMethods = builderMethods.Where(m => !m.ReturnType.IsCollection);
                         conventions.AddRequestBuilderBody(parent, targetClass, writer, prefix: "var builder = ", pathParameters: codeElement.Parameters.Where(x => x.IsOfKind(CodeParameterKind.Path)));
-                        writer.WriteLine("var commands = new List<Command> { ");
+                        writer.WriteLine("var commands = new List<Command> {");
                         writer.IncreaseIndent();
 
-                        foreach (var method in itemBuilderMethods)
+                        foreach (var method in builderMethods.Where(m => !m.ReturnType.IsCollection))
                         {
```

Comment:
filtering o(n) x2, looping o(n)x2
using a single loop with an if inside o(n)

Reply:
Good catch. I'd missed this. Thanks 🙂